presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.plan.ChildReplacer;
+import com.facebook.presto.sql.planner.plan.OutputNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+public class Memo
+{
+    private final PlanNodeIdAllocator idAllocator;
+    private final int rootGroup;
+
+    private int nextGroupId;
+
+    private final Map<Integer, PlanNode> membership = new HashMap<>();
+    private final Map<Integer, Integer> referenceCounts = new HashMap<>();
+
+    public Memo(PlanNodeIdAllocator idAllocator, PlanNode plan)
+    {
+        this.idAllocator = idAllocator;
+        rootGroup = insertRecursive(plan);
+        referenceCounts.put(rootGroup, 1);
+    }
+
+    public int getRootGroup()
+    {
+        return rootGroup;
+    }
+
+    public PlanNode getNode(int group)
+    {
+        checkArgument(membership.containsKey(group), "Invalid group: %s", group);
+        return membership.get(group);
+    }
+
+    public PlanNode extract()
+    {
+        return extract(getNode(rootGroup));
+    }
+
+    private PlanNode extract(PlanNode node)
+    {
+        if (node instanceof GroupReference) {
+            return extract(membership.get(((GroupReference) node).getGroupId()));
+        }
+
+        List<PlanNode> children = node.getSources().stream()
+                .map(this::extract)
+                .collect(Collectors.toList());
+
+        return ChildReplacer.replaceChildren(node, children);
+    }
+
+    public void replace(int group, PlanNode node, String rule)
+    {
+        PlanNode old = membership.get(group);
+
+        // output node is special. TODO: we need to get rid of OutputNode. It doesn't add any value
+        checkArgument(node instanceof OutputNode || new HashSet<>(old.getOutputSymbols()).equals(new HashSet<>(node.getOutputSymbols())),
```

Comment:
Actually, no :)

The way to model PruneXXX rules is to look for, say, a subplan with the following shape:

```
- X
   - Y => [f1, ..., fn]
      - Z
```

Then, determine, whether the columns referenced by expressions in X are a subset of the outputs of Y. If so (say, in this example, only columns f1, f2, f3 are used), rewrite the plan subplan as:

```
- X
   - Y' => [f1, f2, f3]
      - Z
```

With any expression that produces f4, ..., fn in Y removed.

So, X would continue to produce the same outputs. This is a key aspect of the contract for rules and groups in the memo. Rules transform subplans into equivalent subplans, where "equivalent" means, rougly, "they produce the same (multi)set of tuples".